interp/run.go

```diff
@@ -2824,86 +2735,62 @@ func convertConstantValueTo(n *node, typ reflect.Type) {
 	}
 
 	v := n.rval
-	switch typ.Kind() {
+	typ := n.typ.TypeOf()
+	kind := typ.Kind()
+	switch kind {
+	case reflect.Bool:
+		v = reflect.ValueOf(constant.BoolVal(c)).Convert(typ)
+	case reflect.String:
+		v = reflect.ValueOf(constant.StringVal(c)).Convert(typ)
 	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
 		i, _ := constant.Int64Val(c)
 		l := constant.BitLen(c)
-		switch typ.Kind() {
-		case reflect.Int:
-			if l > 64 {
-				panic(fmt.Sprintf("constant %s overflows int", c.ExactString()))
-			}
-			v = reflect.ValueOf(int(i))
-		case reflect.Int8:
-			if l > 8 {
-				panic(fmt.Sprintf("constant %s overflows int8", c.ExactString()))
-			}
-			v = reflect.ValueOf(int8(i))
-		case reflect.Int16:
-			if l > 16 {
-				panic(fmt.Sprintf("constant %s overflows int16", c.ExactString()))
-			}
-			v = reflect.ValueOf(int16(i))
-		case reflect.Int32:
-			if l > 32 {
-				panic(fmt.Sprintf("constant %s overflows int32", c.ExactString()))
-			}
-			v = reflect.ValueOf(int32(i))
-		case reflect.Int64:
-			if l > 64 {
-				panic(fmt.Sprintf("constant %s overflows int64", c.ExactString()))
-			}
-			v = reflect.ValueOf(i)
+		if l > bitlen[kind] {
+			panic(fmt.Sprintf("constant %s overflows int%d", c.ExactString(), bitlen[kind]))
 		}
-	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		v = reflect.ValueOf(i).Convert(typ)
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
 		i, _ := constant.Uint64Val(c)
 		l := constant.BitLen(c)
-		switch typ.Kind() {
-		case reflect.Uint:
-			if l > 64 {
-				panic(fmt.Sprintf("constant %s overflows uint", c.ExactString()))
-			}
-			v = reflect.ValueOf(uint(i))
-		case reflect.Uint8:
-			if l > 8 {
-				panic(fmt.Sprintf("constant %s overflows uint8", c.ExactString()))
-			}
-			v = reflect.ValueOf(uint8(i))
-		case reflect.Uint16:
-			if l > 16 {
-				panic(fmt.Sprintf("constant %s overflows uint16", c.ExactString()))
-			}
-			v = reflect.ValueOf(uint16(i))
-		case reflect.Uint32:
-			if l > 32 {
-				panic(fmt.Sprintf("constant %s overflows uint32", c.ExactString()))
-			}
-			v = reflect.ValueOf(uint32(i))
-		case reflect.Uint64:
-			if l > 64 {
-				panic(fmt.Sprintf("constant %s overflows uint64", c.ExactString()))
-			}
-			v = reflect.ValueOf(i)
-		case reflect.Uintptr:
-			if l > 64 {
-				panic(fmt.Sprintf("constant %s overflows uintptr", c.ExactString()))
-			}
-			v = reflect.ValueOf(i)
+		if l > bitlen[kind] {
+			panic(fmt.Sprintf("constant %s overflows uint%d", c.ExactString(), bitlen[kind]))
 		}
+		v = reflect.ValueOf(i).Convert(typ)
 	case reflect.Float32:
 		f, _ := constant.Float32Val(c)
-		v = reflect.ValueOf(f)
+		v = reflect.ValueOf(f).Convert(typ)
```

Comment:
Do we need the `Convert` here? Also float64 and complex128?

Reply:
I think so. We may have an imported typed constant definition or expression (with type propagation), with an aliased type on `float32`.